### PR TITLE
Start testing Swift 6 mode changes

### PR DESCRIPTION
**Pull request by beccadax**
This pull request makes asserts compilers **only** accept `-swift-version 6` so that we can begin to test source-breaking changes intended for the next language version mode, but without shipping that unfinished work to Swift users. It also adds tests for existing Swift 6 mode behaviors (mostly warnings being strengthened into errors) and enables the changes to `#file` approved in SE-0274/SE-0285 for Swift 6 mode.

Tests that wish to use `-swift-version 6` currently **must** be marked with `REQUIRES: asserts` or they will fail in no-asserts bots. The simplest pattern is probably to duplicate part or all of the Swift 5 test to check that will behave as intended in Swift 6.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci test

File: test/Driver/swift-version-6-noasserts.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Tests temporary -swift-version 6 behavior in compilers with asserts disabled,
+// where we don't allow -swift-version 6 to keep it out of release compilers.
+
+// UNSUPPORTED: asserts
+
+// RUN: not %target-swiftc_driver -swift-version 6 -typecheck %s 2>&1 | %FileCheck --check-prefix ERROR_6 %s
+
+// ERROR_6: <unknown>:0: error: invalid value '6' in '-swift-version 6'
+// ERROR_7: <unknown>:0: note: valid arguments to '-swift-version'
+// ERROR_7-NOT: '6'
```
Comment (kavon) [inline test/Driver/swift-version-6-noasserts.swift:10]: Are we missing a `FileCheck` invocation that uses `ERROR_7` as the prefix?

**Review by kavon — APPROVED** [review]
👍🏼 LGTM. Thanks for doing this and updating the `actor_definite_init` test, among other things.

**Comment by DougGregor** [comment]
I'm going to merge this now so we can start building on it.

**Comment by DougGregor** [comment]
I do wonder if we should use `REQUIRES: swift6` instead to better document the requirement.

**Pull request merged**